bee-message/src/payload/milestone/mod.rs

```diff
@@ -0,0 +1,183 @@
+// Copyright 2020-2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+//! Module describing the milestone payload.
+
+mod essence;
+mod milestone_id;
+
+pub use essence::{MilestonePayloadEssence, MILESTONE_MERKLE_PROOF_LENGTH, MILESTONE_PUBLIC_KEY_LENGTH};
+pub use milestone_id::{MilestoneId, MILESTONE_ID_LENGTH};
+
+use crate::Error;
+
+use bee_common::packable::{Packable, Read, Write};
+
+use crypto::{
+    hashes::{blake2b::Blake2b256, Digest},
+    signatures::ed25519,
+    Error as CryptoError,
+};
+
+use alloc::{boxed::Box, vec::Vec};
+use core::{convert::TryInto, ops::RangeInclusive};
+
+/// Range of allowed milestones signatures key numbers.
+pub const MILESTONE_SIGNATURE_COUNT_RANGE: RangeInclusive<usize> = 1..=255;
+/// Length of a milestone signature.
+pub const MILESTONE_SIGNATURE_LENGTH: usize = 64;
+
+#[derive(Debug)]
+#[allow(missing_docs)]
+pub enum MilestoneValidationError {
+    InvalidMinThreshold,
+    TooFewSignatures(usize, usize),
+    InsufficientApplicablePublicKeys(usize, usize),
+    UnapplicablePublicKey(String),
+    InvalidSignature(usize, String),
+    Crypto(CryptoError),
+}
+
+impl From<CryptoError> for MilestoneValidationError {
+    fn from(error: CryptoError) -> Self {
+        MilestoneValidationError::Crypto(error)
+    }
+}
+
+/// A payload which defines the inclusion set of other messages in the Tangle.
+#[derive(Clone, Debug, Eq, PartialEq)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct MilestonePayload {
+    essence: MilestonePayloadEssence,
+    // TODO Change when serde supports const generics.
+    signatures: Vec<Box<[u8]>>,
+}
+
+impl MilestonePayload {
+    /// The payload kind of a `MilestonePayload`.
+    pub const KIND: u32 = 1;
+
+    /// Creates a new `MilestonePayload`.
+    pub fn new(essence: MilestonePayloadEssence, signatures: Vec<Box<[u8]>>) -> Result<Self, Error> {
+        if !MILESTONE_SIGNATURE_COUNT_RANGE.contains(&signatures.len()) {
+            return Err(Error::MilestoneInvalidSignatureCount(signatures.len()));
+        }
+
+        if essence.public_keys().len() != signatures.len() {
+            return Err(Error::MilestonePublicKeysSignaturesCountMismatch(
+                essence.public_keys().len(),
+                signatures.len(),
+            ));
+        }
+
+        // TODO check signature length
```

Comment:
There are a few TODOs that look a bit more involved, but this seems like a small change.